assert: Add support for architecture specific assert behavior

hersp opened the pull request:
Add new kconfig for making it possible to control the behavior
of an assert depending on platform. If enabled __ARCH_ASSERT must
be defined in architecture to e.g. reset or newer ending loop.

Signed-off-by: Peter Herager <pehe@oticon.com>

hersp commented:
I have added this to be able to be to control the assert behavior, e.g. having a fast reset after the assert. Also this makes it possible to choose to store information about the reset not only having it printed with printk.

galak reviewed (commented):
We'd usually want at least one user of this in tree.

aescolar commented on include/misc/__assert.h:
```diff
@@ -86,11 +86,16 @@
 #else
+#if defined(CONFIG_ASSERT_ARCH_DEFINED)
+  #include "arch/cpu.h"
+  #define __ASSERT_POST __ARCH_ASSERT(__FILE__, __LINE__)
```
Would be logical to have the posix arch use this (see 2 lines above)

hersp replied:
Agree

andyross reviewed (commented):
No complaints other than terminology: this isn't an architecture defined "assert", it's an architecture-defined post-assert hook to replace the infinite loop we currently have.

Also I'm not sure if this really wants to be architecture-specific at all.  It's totally reasonable to imagine an application wanting to do its own hooks irrespective of whatever it got from its arch layer.

Maybe make a C API instead of a preprocessor one?  So you could e.g. set CONFIG_POST_ASSERT_HANDLER to the symbol name of a function and default it to something appropriate in your arch?  Or you could even make it a runtime API where you could register a callback, though that seems sorta silly for this use case.

Finally note that there is a patch floating around somewhere that wants to make asserts call k_panic (@pfalcon, yours I think?), which would slot right into a design like this too.

hersp commented:
I do not think a function interface would be correct here. At least from my perspective it should be possible to record the address of the assert in an easy way(a system with restricted amounts of memory would record this not the file name), a function call would make this more difficult in some architectures.

andyross commented:
_ASSERT() already reads the file and line number.  There's no reason it couldn't provide this information to your handler.  I guess I don't know exactly what you want your handlers to do.  File/line/function metadata like that isn't really arch-specific and would be useful to everyone, right?  So add that stuff to the general API and put the low level behvaior (panic, halt-and-catch-fire, whatever) into the new API.

hersp commented:
What I want is not the file and line number, but the program/instruction counter, as this takes up less space than the file name+line number. I do not want strings that are really not needed. I I see it this cannot easily be handled by a function call but with a macro it is not an issue.

hersp commented:
Committed new version with naming changed to post assert and now used by the posix arch.

aescolar commented on subsys/debug/Kconfig:
```diff
@@ -159,6 +159,14 @@ config FORCE_NO_ASSERT
 	  CFLAGS and not Kconfig.  Added solely to be able to work
 	  around compiler bugs for specific tests.
 
+config CONFIG_POST_ASSERT_ARCH_DEFINED
```
config POST_ASSERT_ARCH_DEFINED
without the "CONFIG_"

aescolar commented on subsys/debug/Kconfig:
```diff
@@ -159,6 +159,14 @@ config FORCE_NO_ASSERT
+    default y if ARCH_POSIX
+	help
+	  This boolean option enables architecure specific implemenation of
+	  post assertion behavior. Added to let architecture define if and how
```
Nit: "if and how to reset when an assert occurs." Implies the arch would want to reset. Maybe you want to say "what to do after an asserts occurs"

hersp committed:
assert: Add support for architecture specific post assert behavior

Add new kconfig option for making it possible to control the post assert
behavior depending on platform. If enabled __ARCH_POST_ASSERT must be
defined in the architecture to for ex. reset the platform.

Signed-off-by: Peter Herager <pehe@oticon.com>

aescolar reviewed (approved):
Seems good enough to me, 
@andyross @andrewboie : Are you happy enough with it?

tyrantlizard77 reviewed (changes requested):
I'd rather this was not merged.
This is not a policy decision to make on a per-arch basis.
What __ASSERT() really should be doing is call k_panic().
Then in _SysFatalErrorHandler(), which can be overridden with a custom implementation at the application level, you can decide what the policy is when the reason code is _NANO_ERR_KERNEL_PANIC.

hersp commented:
I can vote for the k_panic solution, and change the PR to this solution. This is also a cleaner solution as we do not introduce a new option but call k_panic every time. Only minor is that we get two printk containing same filename, and line.

andrewboie commented:
This has been done, see #12732

aescolar commented:
Closing, as #12732 seems to cover @hersp needs and conflicts with this solution (if you disagree feel free to reopen or create a new PR)
